src/dataflow/DataflowAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,379 @@
+        TypeElement typeElt = types.boxedClass(type.getUnderlyingType());
+        AnnotationMirror am = DataflowUtils.createDataflowAnnotation(typeElt.asType().toString(),
+                this.processingEnv);
+        AnnotatedDeclaredType dt = fromElement(typeElt);
+        dt.addAnnotation(am);
+        return dt;
+    }
+
+    /**
+     * This method handles unboxing for reference type.
+     * For statements, int i = new Integer(3);
+     * The annotation for i should be @DataFlow(typeNames = {"int"}).
+     */
+    @Override
+    public AnnotatedPrimitiveType getUnboxedType(AnnotatedDeclaredType type)
+            throws IllegalArgumentException {
+        PrimitiveType primitiveType = types.unboxedType(type.getUnderlyingType());
+        AnnotationMirror am = DataflowUtils.createDataflowAnnotation(primitiveType.toString(),
+                this.processingEnv);
+        AnnotatedPrimitiveType pt = (AnnotatedPrimitiveType) AnnotatedTypeMirror.createType(
+                primitiveType, this, false);
+        pt.addAnnotation(am);
+        return pt;
+    }
+
+    private final class DataFlowQualifierHierarchy extends GraphQualifierHierarchy {
+
+        public DataFlowQualifierHierarchy(MultiGraphFactory f, AnnotationMirror bottom) {
+            super(f, bottom);
+        }
+
+        /**
+         * This method checks whether rhs is subtype of lhs. rhs and lhs are
+         * both Dataflow types with typeNameRoots argument.
+         * 
+         * @param rhs
+         * @param lhs
+         * @return true is rhs is subtype of lhs, otherwise return false.
+         */
+        private boolean isSubtypeWithRoots(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            Set<String> rTypeNamesSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNames(rhs)));
+            Set<String> lTypeNamesSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNames(lhs)));
+            Set<String> rRootsSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNameRoots(rhs)));
+            Set<String> lRootsSet = new HashSet<String>(Arrays.asList(DataflowUtils
+                    .getTypeNameRoots(lhs)));
+            Set<String> combinedTypeNames = new HashSet<String>();
+            combinedTypeNames.addAll(rTypeNamesSet);
+            combinedTypeNames.addAll(lTypeNamesSet);
+            Set<String> combinedRoots = new HashSet<String>();
+            combinedRoots.addAll(rRootsSet);
+            combinedRoots.addAll(lRootsSet);
+
+            AnnotationMirror combinedAnno = DataflowUtils.createDataflowAnnotationWithRoots(
+                    combinedTypeNames, combinedRoots, processingEnv);
+            AnnotationMirror refinedCombinedAnno = refineDataflow(combinedAnno);
+            AnnotationMirror refinedLhs = refineDataflow(lhs);
+
+            if (AnnotationUtils.areSame(refinedCombinedAnno, refinedLhs)) {
```

Comment:
1. Dataflow arguments are connected by disjunction rather than conjunction. You can see each argument is independent to each other. `@Dataflow(roots={Number}, names={Object})` is valid, and we can read it as the annotated type is either any subtype of Number or the Object type. It would happen in following code:
```java
Object m() {
    if (...) {
        return new Object();
    } else {
        return m2(); // m2 is from byte code, and the return type of it is Number.
    }
}
```
The line in my thesis you mentioned may be confusing. "all possible upper bounds of Java run-time types a method may return", it's not all possible upper bounds, it's the all possible upper bounds we can observe, and those upper bounds and the arguments in `typeName` are independent to each other.

2. The algorithm there is not simply checking the superset relation, instead, it checks whether T1 can cover more types than T2.
The algorithm works as follows:
    1. Simplify T1 and T2 (There is a section in my thesis discusses this simplification process)
    2. Concatenate T1 and T2. (Results `combinedAnno` variable in line 141).
    3. Simplify `combinedAnno`.
    4. If `combinedAnno` = T2, then T2 is supertype of T1, vice versa.

So if we run this algorithm on your example: T1 = `@Dataflow(roots=Object)`, T2 = `@Dataflow(roots="", names="String")`
    1. There is nothing to simplify on both T1 and T2. Skip this step.
    2. `combinedAnno = @Dataflow(roots = "Object", names = "String")`.
    3. Simplify `combinedAnno`, and the result is `@Dataflow(roots = "Object")`. 
    4. Simplified `combineAnno` = T1, so T1 is supertype of T2.